### ISSUE TITLE
Add support for nested settings

### ISSUE THREAD

**Issue by stinovlas** [Genida/django-appsettings#33]
It would help the readability if `django-appsettings` supported nested settings out of the box. Consider, that I have complicated app with lots of settings and I want to split these settings into groups like this:

```python
MYAPP_API = {
    'SERVER': 'localhost',
    'PORT': 80,
    'SOME_LIST_SETTING': [...],
    ...
}
```

I'd like to define that I have nested setting `MYAPP_API` which consists of several other settings (in this case `StringSetting server`, `IntegerSetting port` and `ListSetting some_list_setting`. Each of these subsettings should be validated separately. Not all of them has to be present, depending on `required` param for each of these subsettings. My settings could then look like this:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        settings={
            'server': appsettings.StringSetting(required=True),
            'port': appsettings.IntegerSetting(default=80),
            'username': appsettings.StringSetting(),
            ...
        }
    )

    class Meta:
        setting_prefix = 'myapp_'
```

Checker would then validate each inner setting and the nested setting itself (it would be required if and only if any of it's inner settings is required). I don't know whether there are any relevant arguments that you may want to pass to the `NestedSetting`. If there aren't, we could even use pretty kwargs syntax:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
            server=appsettings.StringSetting(required=True),
            port=appsettings.IntegerSetting(default=80),
            username=appsettings.StringSetting(),
            ...
        }
    )

    class Meta:
        setting_prefix = 'myapp_'
```

**Comment by stinovlas** [comment]
I'd appreciate some help in designing the interface, but other than that, I'm willing to implement this.

**Comment by pawamoy** [comment]
Just had an idea: couldn't we nest the `AppSettings` class directly instead? Something like the following:

```python
import appsettings

class ApiSettings(appsettings.AppSettings):
    server = appsettings.StringSetting(required=True)
    port = appsettings.IntegerSetting(default=80)
    username = appsettings.StringSetting()


class AppConf(appsettings.AppSettings):
    api = ApiSettings()
    # maybe repeat the settings here as well
    # if you want the user to be able to specify them
    # as one single dict or with multiple variables


appconf = AppConf()
```

Because the behavior of the `NestedSetting` is somewhat almost equivalent to what the `AppSettings` class does.

But I don't know if this would be easier (and cleaner) to implement than a `NestedSetting` class.

I'll give you more insights later (what would roughly be needed for both versions).

**Comment by pawamoy** [comment]
So let's say we declare the nested setting like this:
```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        # name is auto-filled,
        # prefix is auto-filled,
        settings=dict(
            server=appsettings.StringSetting(required=True),
            port=appsettings.IntegerSetting(default=80),
            username=appsettings.StringSetting(),
            ...
        )
    )

    class Meta:
        setting_prefix = 'myapp_'


settings = MyAppSettings()
```

How would you access these nested settings now?

1. Like this?
```python
settings.api.server
```
2. Or like that?
```python
settings.api['server']
```


### But I have yet another idea.
Why not directly support the dict format in the `AppSettings` and `Setting` classes?

```python
class MySettings(appsettings.AppSettings):
    server = appsettings.StringSetting(required=True)
    port = appsettings.IntegerSetting(default=80)
    username = appsettings.StringSetting()
    ...

    class Meta:
        setting_prefix = 'myapp_'


# and in settings.py
MYAPP = {
    'SERVER': 'localhost',
    'PORT': 80,
    'SOME_LIST_SETTING': [...],
    ...
}
```

This would immediately allow to use both dict and one-setting-per-variable formats. It should simply be a matter of updating the `raw_value` property of `Setting` to try to fetch the setting from a dict whose name is the setting prefix. I'm not sure how we would handle more nesting (like double/triple nesting), but I feel like it would be possible if we play with the prefixes, splitting them on underscores.

What do you think?

**Comment by ziima** [comment]
> Why not directly support the dict format in the `AppSettings` and `Setting` classes?
> ```python
> # and in settings.py
> MYAPP = {
>     'SERVER': 'localhost',
>     'PORT': 80,
>     'SOME_LIST_SETTING': [...],
>     ...
> }
> ```

Django REST framework has settings in this style, see https://www.django-rest-framework.org/api-guide/settings/, but I wouldn't recommend it. It's really painful if you want to run several processes with slightly different settings. It can be handled in individual settings, such as `TEMPLATES`, but I would really avoid having massive dictionaries for a whole application.

**Comment by stinovlas** [comment]
I'd prefer accessing nested settings with `settings.api.server` notation, but both would work, I guess.

As @ziima pointed out, direct support of the dict format for `AppSettings` would require writing all application settings into single dictionary. `NestedSettings` that I proposed are more refined than that ­– you can define one nested setting for `api`, the other one for `important_links` and several other plain (not-nested) settings as well.

Nesting `AppSettings` class directly is interesting idea, but I realized, that it could be desirable do define some properties for the `NestedSetting` such as `name`, `prefix` or `required` and it would be strange to set these to `AppSettings` objects. These are properties of `Setting` object. Consider following example:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        name='rest_api',
        prefix='my_awesome_',
        required=True,
        settings=dict(
            server=appsettings.StringSetting(required=True),
            port=appsettings.IntegerSetting(default=80),
            username=appsettings.StringSetting(),
        )
    )
    another_setting = appsettings.StringSetting(default='whatever')

    class Meta:
        setting_prefix = 'myapp_'

settings = MyAppSettings()
```

**Comment by stinovlas** [comment]
However, I'm not sure how `required` argument should be handled. Consider this:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        required=False,
        settings=dict(
            server=appsettings.StringSetting(required=True),
        )
    )
    ...
```

1. Should this configuration be invalid? Or should it mean that `MYAPP_API` is optional but if it is defined, it has to contain item `SERVER`? I think invalid configuration would be more appropriate.

1. Should we allow `required` argument at all for `NestedSetting`? If we don't we save ourselves some trouble with the previous question. If we do, then we get the ability to force user to write at least empty dict configuration.

I see two reasonable solutions:

1. Don't allow `required` arg for `NestedSettings`.

1. Allow `required` arg for nested settings and if it is False and nested setting contains item that is `required`, throw exception. If `required` arg is not specified, it is True if and only if any of its nested items is `required`.

Well, when I see the specification, I'd prefer the simple way of not allowing the `required` arg for nested settings.

**Comment by stinovlas** [comment]
And yet, there is another problem of how prefixes should be handled. What if I have this:

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        prefix='outer_prefix_'
        settings=dict(
            server=appsettings.StringSetting(required=True, name='slave',
                                             prefix='inner_prefix_'),
        )
    )

    class Meta:
        setting_prefix = 'myapp_'
```

What would be the appropriate configuation? Perhaps this?

```python
OUTER_PREFIX_API = {
    'INNER_PREFIX_SLAVE': 'localhost',
}
```

Or should we forbid to define prefixes inside nested settings?

**Comment by ziima** [comment]
> However, I'm not sure how required argument should be handled. Consider this:
> ```python
> class MyAppSettings(appsettings.AppSettings):
>     api = appsettings.NestedSetting(
>         required=False,
>         settings=dict(
>             server=appsettings.StringSetting(required=True),
>         )
>     )
> ```
I consider this a perfectly valid setup. Default value for `API` setting is `None` with meaning API is tuned off. But if I want to turn it on, I have to provide `SERVER` subsetting.

**Comment by stinovlas** [comment]
> I consider this a perfectly valid setup. Default value for API setting is None with meaning API is tuned off. But if I want to turn it on, I have to provide SERVER subsetting.

OK, then what about this?

```python
class MyAppSettings(appsettings.AppSettings):
    api = appsettings.NestedSetting(
        required=False,
        default=dict(),
        settings=dict(
            server=appsettings.StringSetting(required=True),
        )
    )
```
Is this invalid setup? And how would relate default values for nested setting and its items?

**Comment by ziima** [comment]
> Is this invalid setup? And how would relate default values for nested setting and its items?

That is a bit tricky. What about something like `empty_values`, which is used in Django form fields?

**Comment by pawamoy** [comment]
I see that we're asking ourselves the same questions :sweat_smile:. The difficult part here is to clearly define our specification. Once it's done, I guess the implementation will not be too complicated.

Just a note about the ideas I wrote above: I agree with both you in saying that they feel too much implicit or counter-intuitive, so I would indeed prefer something more explicit like a `NestedSetting` class.

As we can see, nested settings bring a lot of complexity with how they should behave. I think it could be summarized like that: *how does the nested setting root (`api` in the examples) affect its sub-settings properties?*

- what about `name`? 
- what about `prefix`? should we come up with some prefix concatenation?
- what about `required`? both the examples you gave above feel valid to me. In the end it is the developer who chooses to set the default value to a dict or not. They should then handle that themselves.
- what about `default`? and `call_default` and `transform_default`?
- how the settings are accessed? does `settings.api` return a dict object (the value, cached ot not) or a `NestedSetting` instance with extended fetching capabilities? trying to access `settings.api.server` (with custom `__getattr__`) or `settings.api['server']` (with custom `__getitem__` to handle lowercase) should in both cases raise an `AttributeError` if server was not defined (because it is required).
- how do we handle more nesting? with recursion I guess :grin:

**Comment by stinovlas** [comment]
Here's my proposal:

## Name
The name of the outer setting should not affect the name of the inner setting and vice versa.

## Prefixes
I propose that inner settings will have empty prefix by default (not inherited from `Meta`). However, it would still be possible to define `prefix` arg. Prefixes would not concatenate, because they are already present in the name of the enveloping dictionaries. Example:

```python
class MyAppSettings(appsettings.AppSettings:
    api = appsettings.NestedSetting(
        prefix='our_',
        settings=dict(
            server = appsettings.StringSetting(prefix='my_')
            port = appsettings.IntegerSetting()
        )
    )

# in settings.py:
OUR_API = {
    'MY_SERVER': ...,
    'PORT': ...
}
```

## Defaults
If the outer setting is not set, it's default value should be used, no matter what the default values of inner settings are. You could even define outer default value that is incompatible with inner settings types. You are allowed to shoot yourself in the leg if you wish so.

If outer setting is set, it's default value does not apply. Inner settings default values apply, if the appropriate inner setting is omitted. 

The same applies for call_default and transform_default. Example:

```python
class MyAppSettings(appsettings.AppSettings):
    alphabet = appsettings.NestedSetting(
        default=dict(a=1, b=2),
        settings=dict(
            a=appsettings.IntegerSetting(default=5),
            b=appsettings.IntegerSetting(default=6),
        )
    )

settings = MyAppSettings()
```
```python
# empty settings.py
# settings.alphabet.a == 1
# settings.alphabet.b == 2
```
```python
# in settings.py:
MYAPP_ALPHABET = {}
# settings.alphabet.a == 5
# settings.alphabet.b == 6
```
```python
# in settings.py:
MYAPP_ALPHABET = {'A': 3}
# settings.alphabet.a == 3
# settings.alphabet.b == 6
```

## Required
Basically, the same logic as with defaults applies. If outer setting is not required, it doesn't have to be set (no matter if the inner settings are required). If the outer setting is set, all the inner settings that are required must be present. Example:

```python
class MyAppSettings(appsettings.AppSettings):
    alphabet = appsettings.NestedSetting(
        default=None,
        required=False,
        settings=dict(
            a=appsettings.IntegerSetting(required=True),
            b=appsettings.IntegerSetting(default=6),
        )
    )

settings = MyAppSettings()
```
VALID SETTINGS:
* empty settings
* `MYAPP_ALPHABET = {'A': 5}`
* `MYAPP_ALPHABET = {'A': 5, 'B': 6}`

INVALID SETTINGS:
* `MYAPP_ALPHABET = {}`
* `MYAPP_ALPHABET = {'B': 6}`

**Comment by stinovlas** [comment]
I think I could sum the main thought into two simple rules:
1. `NestedSetting` is a setting.
1. Arguments of the setting affect only the setting itself, not it's ancestors nor descendants.

**Comment by ziima** [comment]
> * how the settings are accessed? does `settings.api` return a dict object (the value, cached ot not) or a `NestedSetting` instance with extended fetching capabilities? trying to access `settings.api.server` (with custom `__getattr__`) or `settings.api['server']` (with custom `__getitem__` to handle lowercase) should in both cases raise an `AttributeError` if server was not defined (because it is required).

I strongly support the dictionary API. As I noted above - if it's dictionary in settings module, it should behave as such in code. Therefore I suggest to
 * return dictionary when the master setting, e.g. `settings.api`, is accessed
 * use item getters to access the individual members

I'm not quite i favor of using the lowercase when referencing the settings, as the common usage in Django and various Django applications is the uppercase, but I wouldn't mind that much if the dictionary would be case-insensitive. We can probably copy it from `requests`, where it's used for dictionary with HTTP headers.

**Comment by ziima** [comment]
Otherwise +1 for @stinovlas 's proposal

**Comment by stinovlas** [comment]
I'm going on a vacation for a week or so, but when I return, I'd like to finish the discussion about the API. In general, I'm OK with dictionary API.

**Comment by pawamoy** [comment]
Thank you @stinovlas, great proposal :+1:! I agree with all the points you wrote. I just have one more remark below :mag_right:  

@ziima, about using lowercase when accessing the settings: I agree that nobody should be allowed to write the project's settings **variables** in lowercase. So in our case, user must write `API = {'SERVER': ...}` and not `API = {'server': ...}`. But in the `MyAppSettings` class, the server setting was declared as `server = ...`. My question/concern here is: which one do we use to access the setting value?

Let's take back the prefix example from @stinovlas:

```python
class MyAppSettings(appsettings.AppSettings:
    api = appsettings.NestedSetting(
        prefix='our_',
        settings=dict(
            server = appsettings.StringSetting(prefix='my_')
            port = appsettings.IntegerSetting()
        )
    )

# in settings.py:
OUR_API = {
    'MY_SERVER': ...,
    'PORT': ...
}
```

Would we access the server setting with:
1. `settings.api['MY_SERVER']` (setting's full name: uppercase prefix + uppercase name),
2. `settings.api['SERVER']` (setting's uppercase name),
3. or `settings.api['server']` (setting's name)

Said differently: would `settings.api` return `{'MY_SERVER': ...}` or `{'SERVER': ...}`?

It's also related to what @ziima told with:
> - return dictionary when the master setting, e.g. settings.api, is accessed
> - use item getters to access the individual members

**Comment by ziima** [comment]
We had a discussion about lowercase vs. uppercase with @stinovlas and despite his dislike we agreed on using uppercase. In probably all projects I've seen and developed all settings are typed in uppercase. In projects where I use appsettings I define the settings and prefixes also in uppercase.
Hence the correct usage in my opinion is `SETTINGS.API['SERVER']`.

I would expect the `SETTINGS.API` to return `{'SERVER': ...}`. As I understand, the prefix serves to differentiate the name in settings module and in code.

**Comment by stinovlas** [comment]
@pawamoy: I think we should access the server setting with `settings.api['server']`. Name of the item should correspond with name of the settings class variable. If you want to use uppercase everywhere, you can do it like this:

```python
class MyAppSettings(appsettings.AppSettings):
    API = appsettings.NestedSetting(
        settings=dict(
            SERVER = appsettings.StringSetting()
        )
    )

SETTINGS = MyAppSettings()
```

Then you can access your server settings with `SETTINGS.API['SERVER']`. Just let the user choose whether he wants to use uppercase everywhere, or not.

**Comment by pawamoy** [comment]
I'm on a vacation as well now, coming back thursday 19th!

**Comment by pawamoy** [comment]
I'm back. @stinovlas you still want to implement this? I think we have a good idea of how this feature should behave. I am sure you will still have some surprises when actually implementing it though.

I often feel like this application is a good case for type hints ([pydantic?](https://github.com/samuelcolvin/pydantic)) or something equivalent, and that we try to reinvent the wheel in a weird manner :confused:, but anyway, let me hear your thoughts!

**Comment by stinovlas** [comment]
@pawamoy: Yes, I want to do that. Hopefully, I'll get to it during this week.
